src/ppx/instrument.ml

```diff
@@ -1482,75 +1470,58 @@ class instrumenter =
              inside [struct]..[end] or in an attribute or extension point.
              Traverse the structure recursively as normal. *)
+      structure_instrumentation_suppressed <-
+        saved_structure_instrumentation_suppressed;
 
-        else begin
-          (* This is the first structure we see in te file, and we are not in an
-             interface file, so Bisect_ppx is beginning to (potentially)
-             instrument the current file. We need to check whether this file is
-             excluded from instrumentation before proceeding. *)
-          saw_top_level_structure_or_signature <- true;
-
-          let path = !Location.input_name in
-
-          let file_should_not_be_instrumented =
-            (* Bisect_ppx is hardcoded to ignore files with certain names. If we
-               have one of these, return the AST uninstrumented. In particular,
-               do not recurse into it. *)
-            let always_ignore_paths = ["//toplevel//"; "(stdin)"] in
-            let always_ignore_basenames = [".ocamlinit"; "topfind"] in
-
-            List.mem path always_ignore_paths ||
-            List.mem (Filename.basename path) always_ignore_basenames ||
-            Exclusions.contains_file path ||
-            Coverage_attributes.has_exclude_file_attribute ast
-          in
+      result
 
-          if file_should_not_be_instrumented then
-            ast
-
-          else begin
-            (* This file should be instrumented. Traverse the AST recursively,
-               then prepend some generated code for initializing the Bisect_ppx
-               runtime and telling it about the instrumentation points in this
-               file. *)
-            let instrumented_ast = super#structure ast in
-            let runtime_initialization =
-              Generated_code.runtime_initialization points path
-            in
-            runtime_initialization @ instrumented_ast
-          end
-        end
+    method transform_impl_file ast =
+      let saved_structure_instrumentation_suppressed =
+        structure_instrumentation_suppressed in
+
+      let result =
+        let file_should_not_be_instrumented path =
+          (* Bisect_ppx is hardcoded to ignore files with certain names. If we
+            have one of these, return the AST uninstrumented. In particular,
+            do not recurse into it. *)
+          let always_ignore_paths = ["//toplevel//"; "(stdin)"] in
+          let always_ignore_basenames = [".ocamlinit"; "topfind"] in
+
+          List.mem path always_ignore_paths ||
+          List.mem (Filename.basename path) always_ignore_basenames ||
+          Exclusions.contains_file path ||
+          Coverage_attributes.has_exclude_file_attribute ast
+        in
+
+        let get_path = function
+          | [] -> None
+          | hd :: _ -> Some hd.pstr_loc.loc_start.pos_fname
+        in
+
+        match get_path ast with
+          | None -> ast
```

Comment:
Under what condition(s) will this be `None`? I suspect that we want to instrument in those cases, if they are possible.

Reply:
This will be `None` whenever the structure given to the rewriter is empty. Two cases I know of in which the structure is empty:
- When having an empty .ml file, the structure corresponding to that file is empty.
- When loading a toplevel loading a ppx deriver (as done for example in `test_top_level` test), the deriver will get loaded and applied to an empty structure before getting loaded and getting applied to the input. To be honest, I don't know why it's like that.

Reply:
Is it ok to leave it the way it is? Or would you like me to find out all possible cases in which the structure is empty?

Reply:
I've recently understood that the way I was trying to get the source filename might be problematic, since `hd.pstr_loc` might be `Location.none`, for example if a ppx runs before `bisect_ppx` using `Location.none`.  I've also just seen that the source filename is not only used to decide if to traverse the AST or not, but also inside the runtime initialization. So it doesn't seem like a good idea to try to get around using `Location.input_name` here. I've gone back to using it, even though that means relying on compiler-libs here. If ppxlib introduces a similar concept in the near future or I find another way around it, I'll adapt the PR.